architecture_records/004-dataloader-v2.md

```diff
@@ -0,0 +1,314 @@
+# Data Loader Design For FMS-HF-Tuning
+
+**Deciders(s)**: Sukriti Sharma (sukriti.sharma4@ibm.com), Will Johnson (Will.Johnson@ibm.com) , Abhishek Maurya (maurya.abhishek@ibm.com), Dushyant Behl (dushyantbehl@in.ibm.com), Ashok Pon Kumar (ashokponkumar@in.ibm.com)
+
+**Date (YYYY-MM-DD)**:  2024-03-06
+
+**Obsoletes ADRs**:  NA
+
+**Modified By ADRs**:  NA
+
+**Relevant Issues**: [1]
+
+- [Summary and Objective](#summary-and-objective)
+  - [Motivation](#motivation)
+  - [User Benefit](#user-benefit)
+- [Decision](#decision)
+  - [Alternatives Considered](#alternatives-considered)
+- [Consequences](#consequences)
+- [Detailed Design](#detailed-design)
+
+## Summary and Objective
+
+The reason for motivating dataloader design for fms-hf-tuning is to have a unified interface which supports many type of data formats, streaming and non streaming data, weight based data mixing and many others. Full list we focus on is below - 
+
+1. Data Loading -
+
+    1. Different formats of data → Arrow, Parquet etc.
+
+    1. Different modalities of data -> Images, Audio etc.
+
+    1. Streaming data.
+
+    1. Data Replay
+
+    1. Resume with different number of GPUs
+
+    1. Async data loading
+
+1. Data Preprocessing - 
+
+    1. Custom Attn Masking
+
+    1. Tool Usage
+
+1. Data Mixing - 
+
+    1. Static weights based mixing
+
+### Motivation
+
+### User Benefit
+
+## Decision
+
+### Alternatives Considered
+
+## Consequences
+
+### Advantages
+
+### Impact on performance
+
+## Detailed Design
+
+The input spec which user specifies on how to pass information to such data loader is this
+
+```
+dataloader:
+    type: default(hfdataloader)/stateful(fms-fsdp implementation)
+    streaming: true
+datasets:
+  - name: dataset1
+    sampling:
+      ratio: 0.3
```

Comment:
@dushyantbehl @Ssukriti I feel the better design is not to expose so many fine grained configurations to the user. Consider the the typical use case for loading datasets

```python
from datasets import load_dataset
data = load_dataset("alpaca")
```

When users load a dataset, they just want to use it, they dont really want to care to set complicated flags like "streaming: True". 

In my experience, dataset preperation is very specialized skill. If you think about ILAB, the data is prepared  by the Cambridge team and we just take the tokens and train with it. As such, a recipe is prepared and given to regular users who just deal with the preprocessed data. Simple things like choosing the wrong chat template can severely degrade performance, so regular users wont even mess with that.

Thus in fms-hf-tuning, there is no way to pass such a reciepe, because users will have to set their own `response_template`, etc. I feel that this ADR should aim to provide these reciepes

However what this ADR is doing now is focusing a lot on these fine-detailed configs, which by my argument, a regular user will not mess with. I think we can take inspiration with how HF does these things with fine-grained configs 

```python
from transformers import AutoModelForCausaLMl
model = transformers.AutoModelForCausalLM("....")
```
this actually downloads a `config.json` with complicated model settings that users dont really look at, but is very important to the ensuing that the model is correctly built.

So then the analogue that we might want is

```python
from fms_datasets import load_datasets

# train_dataset is a torch.Dataset that can just be passed to trainer, it
# should be already processed (or can be processed on the fly)
train_dataset = load_datasets('....")

# if train_dataset is multimodal, then some things like
# tokenizer.apply_chat_template(train_dataset, tokenize=False) may not make sense anymore
# so this has to be considered in the ADR
```
where you point to a path with a fine-grained config, and it does the data prep completely for the user. 

Therefore with this abstraction, someone well-versed in dataset prep, will then 
- prepare and publish the fine-grained configs, and then pass them to regular users to be used
- the publisher of such configs, could be IBM, or could be power-users in the public 

Some thoughts:
- performance is key. The more configs you give to the user, the more you need to think about how to store the intermediate results. If a dataset is not tokenized, then the tokenizer needs to run every time training starts. HF datasets caches this automatically. However, if chat templates are providable, and dataset mixing is allowable, then how are all of these going to be cached?
- If a dataset is big, then it should be handled automatically for the user without a user needing to set `streaming:true`
- If its multimodal the mixing should also be handling for the user, maybe allowing the user to select some high-level parameters like mixing constants
- chat templates, response templates, etc, any preprocessing config should be just written out in the fine-grained config. This is along the lines of "code-as-documentation".
- we should search out other open source repos that does the simimilar thing, and discuss why we cant use them
- we should call out why a standard HF dataset is insufficient, and why we need to extend (to a `DataLoader` as you call it)
- the ADR should call out high-level goals of the design, like "there is a need to distributed preprocessing recipes. In the "Summary and Objectives" when things like "support Arrow formats" is listed out -> it shoudl be clarified that this is not because HF datasets does not support arrow formats, but t is because fms-hf-tuning forces users to provide json datasets (i know this is kind of implied very late in the ADR).
- I feel we should not invent new abstractions, if a `DataLoader`, as you call it, can be abstracted by a `DataSet`, then its better to go with the latter, since everybody is familiar with it. Less abstractions, the more unstable the code.
- the concepts of `DataHandlerConfig, DataConfig, DataLoaderConfig` are completely internal, and to some extent, is of lesser importance to the ADR, whereas the higher level concerns, like usability and architectural considerations, need to be first ironed out and clarified before going into internals.

Reply:
I agree we need to talk about design more. I have expressed my dissatisfaction with it in some issues as well. We can discuss on a call next week

Reply:
@ashokponkumar can we please move away from providing default configs to auto detection for standard use cases? Providing default configs is still very messy and confusing . 

- for advanced users it is fine to allow to customize, but for straightforward use cases , we need to shortlist what data formats we want to support in repo - **accept datasets as one argument -> call preprocessing and training as needed** 
- for users that want to pass multiple datasets and need sampling , they can use the config file, but again specifying data preprocessing function in the config need not be compulsory?